### PR TITLE
:sparkles: Add bit_bang_spi

### LINKED ISSUE #17: Add `hal::soft::bit_bang_spi`

### PR DESCRIPTION
**Pull request by MaliaLabor**
Resolves #17

### CODE REVIEW

File: include/libhal-soft/bit_bang_spi.hpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+   * @param p_pins named structure that contains pointers to the sck, copi, and
+   * cipo to be used inside of the driver
+   * @param p_steady_clock the steady clock that will be used for timing the sck
+   * line.
+   * @param p_settings the initial settings of the spi bus
+   */
+  bit_bang_spi(pins const& p_pins,
+               steady_clock& p_steady_clock,
+               settings const& p_settings = {},
+               delay_mode p_delay_mode = delay_mode::with);
```
Comment (kammce) [inline include/libhal-soft/bit_bang_spi.hpp:62]: Missing docs for p_delay_mod

File: include/libhal-soft/bit_bang_spi.hpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    output_pin* sck;
+    output_pin* copi;
+    input_pin* cipo;
+  };
+
+  enum class delay_mode
+  {
+    with,
+    without
+  };
```
Comment (kammce) [inline include/libhal-soft/bit_bang_spi.hpp]: Can you add comments here to describe what happens when these are used at construction.

File: include/libhal-soft/bit_bang_spi.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  bool m_phase;
+
+  /// @brief Time of a full read/write cycle, used in delay_mode::with
+  hal::time_duration m_cycle_duration;
+
+  /// @brief Configuration for using delays or not
+  delay_mode m_delay_mode;
+};
+
+}  // namespace hal::soft
```
Comment (kammce) [inline include/libhal-soft/bit_bang_spi.hpp]: Add newline at the end of the file.

File: src/bit_bang_spi.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  } else {
+    // when phase is 0 (false), data is written on the leading edge of the clock
+    // and data is read in on the falling edge
+    m_copi->level(p_bit_to_write);
+    m_sck->level(!m_polarity);
+    return m_cipo->level();
+  }
+}
+
+}  // namespace hal::soft
```
Comment (kammce) [inline src/bit_bang_spi.cpp]: Add newline to the file.

**Review by kammce — COMMENTED** [review]
Looks great! Make these changes to the PR in a new commit and we should be good to approve.

File: include/libhal-soft/bit_bang_spi.hpp (hunk 1 above)
Comment (MaliaLabor) [reply include/libhal-soft/bit_bang_spi.hpp:62]: Done, unless the wording isn't what you had in mind

File: include/libhal-soft/bit_bang_spi.hpp (hunk 2 above)
Comment (MaliaLabor) [reply include/libhal-soft/bit_bang_spi.hpp]: Done, unless the wording isn't what you had in mind

File: src/bit_bang_spi.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
+#include <chrono>
+
+#include <libhal-soft/bit_bang_spi.hpp>
+#include <libhal-util/bit.hpp>
+#include <libhal-util/steady_clock.hpp>
+#include <libhal/steady_clock.hpp>
+#include <libhal/units.hpp>
+#include <sys/types.h>
```
Comment (kammce) [inline src/bit_bang_spi.cpp]:
```suggestion
```

Did you need this?

**Review by kammce — APPROVED** [review]
Looks great! Looks like there is a stray include in there.

File: src/bit_bang_spi.cpp (hunk 5 above)
Comment (MaliaLabor) [reply src/bit_bang_spi.cpp]: don't think so, I took it out

**Pull request merged**